### DISCUSSION TITLE
Brew/curl madness on mac 10.10.5

### DISCUSSION THREAD

**Discussion by MrAtheist** [Homebrew/discussions#2265] [Everyday usage]
I have this hand-me-down macbook running 10.10.5, which looks like brew doesnt support anymore. With that said, would someone nudge me in the direction of how to upgrade curl in the first place so that it doesnt complain about the handshake error everywhere? (is there a readme somewhere specifically for 10.10.5?)

```
$ brew upgrade (or doing anything with brew...)
....
.....
curl: (35) SSL peer handshake failed, the server most likely requires a client certificate to connect
Error: Failed to download resource "git--html"
Download failed: https://mirrors.edge.kernel.org/pub/software/scm/git/git-htmldocs-2.33.0.tar.xz
Error: 'git' must be installed and in your PATH!
```

I ended up installing curl 7.79.1 from the [source](https://curl.se/download.html), setting up $PATH, and i still have no idea how to make brew using the new curl. Anyone know any other magic that I can give it a shot?

```
$ ./configure --with-secure-transport
$ make && make install

$ which curl
/usr/local/bin/curl

$ curl --version
curl 7.79.1 (x86_64-apple-darwin14.5.0) libcurl/7.79.1 SecureTransport zlib/1.2.5
Release-Date: 2021-09-22
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS HSTS IPv6 Largefile libz NTLM NTLM_WB SSL UnixSockets

$ export PATH=/usr/local/bin/curl:/usr/local/sbin:/usr/local/bin:$PATH
```

```
$ brew config
HOMEBREW_VERSION: 3.2.8
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: f026dd21c195952747395ebdda24327204f012ec
Last commit: 8 weeks ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 835c842dfb3c668a39e5068ccd2d836936fe7f0b
Core tap last commit: 7 weeks ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: quad-core 64-bit haswell
Clang: 7.0.2 build 700
Git: 2.23.0 => /usr/local/bin/git
Curl: 7.43.0 => /usr/bin/curl
macOS: 10.10.5-x86_64
CLT: 7.2.0.0.1.1447826929
Xcode: 7.2.1
```

**Comment by SMillerDev** [comment]
You can try [`HOMEBREW_FORCE_BREWED_CURL`](https://docs.brew.sh/Manpage), it's not exactly what you asked but it should work for what you want.

**Reply by carlocab** [reply]
This will install Homebrew `curl` for you automatically. You'll encounter errors from installing `curl` because of the `curl` you installed into `/usr/local/bin`. Just do `brew link --overwrite curl` when that happens. (Or you can move your own `curl` out of the way before doing anything.)

**Reply by MrAtheist** [reply]
> This will install Homebrew curl for you automatically.

What's the exact cmd?

With `HOMEBREW_FORCE_BREWED_CURL`, the readme states that...

> If set, always use a Homebrew-installed curl(1) rather than the system version. Automatically set if the system version of curl is too old.

The thing is that i dont have curl installed via brew as it got stuck with handshake error in the first place. I do have my own curl under `/usr/local/bin/`, but brew doesnt recognize it. It seems like im stuck in this infinite loop?

```
$ HOMEBREW_FORCE_BREWED_CURL=1 brew install curl
Warning: You are using macOS 10.10.
We (and Apple) do not provide support for this old version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
old version.

==> Downloading https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz

curl: (35) SSL peer handshake failed, the server most likely requires a client certificate to connect
Error: Failed to download resource "pkg-config"
Download failed: https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz
Error: 'curl' must be installed and in your PATH!
Warning: You are using macOS 10.10.
We (and Apple) do not provide support for this old version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
old version.

==> Downloading https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz

curl: (35) SSL peer handshake failed, the server most likely requires a client certificate to connect
Error: Failed to download resource "pkg-config"
Download failed: https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz


$ HOMEBREW_FORCE_BREWED_CURL=1 brew link --overwrite curl
Warning: Refusing to link macOS provided/shadowed software: curl
```

**Comment by Bo98** [comment]
```
HOMEBREW_VERSION: 3.2.8
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: f026dd21c195952747395ebdda24327204f012ec
Last commit: 8 weeks ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 835c842dfb3c668a39e5068ccd2d836936fe7f0b
Core tap last commit: 7 weeks ago
```

This is outdated - various curl issues on older macOS were fixed in 3.2.15. What happens when you run `brew update`?

It will try to install brewed curl after updating, but if it's all up-to-date then it should work (though installation will take some time). We put the effort in to provide mirrors for everything last week to avoid SSL issues.

**Reply by MrAtheist** [reply]
Always got stuck in this ssl loop, not exactly sure how to proceed... should i just nuke everything and start brew from scratch?

```
$ brew update
Warning: You are using macOS 10.10.
We (and Apple) do not provide support for this old version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
old version.

==> Downloading https://ftp.gnu.org/gnu/gettext/gettext-0.21.tar.xz
Already downloaded: /Users/MrAtheist/Library/Caches/Homebrew/downloads/fbd8967fd8256b2e1856b74e8ff0f3d8371cb652c5768e47649613a69d4dc841--gettext-0.21.tar.xz
==> Downloading https://ftp.pcre.org/pub/pcre/pcre2-10.37.tar.bz2
Already downloaded: /Users/MrAtheist/Library/Caches/Homebrew/downloads/aca498bd073103667423550d9baecd164409be9e50cfcc60a82205963f592fab--pcre2-10.37.tar.bz2
==> Downloading https://mirrors.edge.kernel.org/pub/software/scm/git/git-htmldocs-2.33.0.tar.xz

curl: (35) SSL peer handshake failed, the server most likely requires a client certificate to connect
Error: Failed to download resource "git--html"
Download failed: https://mirrors.edge.kernel.org/pub/software/scm/git/git-htmldocs-2.33.0.tar.xz
Error: 'git' must be installed and in your PATH!
```

**Reply by Bo98** [reply]
Ah, it seems you must have not had Homebrew's git installed before - which is a little surprising since that has been a requirement on 10.10 and 10.11 for a while.

Try `brew update-reset` first and see if `brew update` works then.

**Reply by MrAtheist** [reply]
👍  it's going! I see that it's downloading from mirror when failing to handshake. I'll report back once `brew update` is done.

**Reply by MrAtheist** [reply]
After a bit of messing around, it's not quite there yet but i guess it's better than before...

```
$ brew update-reset
# ...a bunch of git pull... 👍 

$ brew update   
# ... took 30+ min, but im not complaining...

$ git -C /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core fetch --unshallow   
# ... took another 5min, so far so good...

$ brew upgrade
....
....
==> Upgrading node
  9.8.0 -> 16.11.0 
==> Installing dependencies for node: mpdecimal, readline, sqlite, python@3.10, sphinx-doc, libuv and python@3.9
==> Installing node dependency: mpdecimal
==> ./configure --prefix=/usr/local/Cellar/mpdecimal/2.5.1
==> make
Last 15 lines from /Users/MrAtheist/Library/Logs/Homebrew/mpdecimal/02.make:
./decimal.hh:228:8: error: thread-local storage is not supported for the current target
extern thread_local Context context;
       ^
decimal.cc:181:1: error: thread-local storage is not supported for the current target
thread_local Context context{context_template};
^
decimal.cc:181:1: error: thread-local storage is not supported for the current target
thread_local Context context{context_template};
^
2 errors generated.
2 errors generated.
make[1]: *** [.objs/decimal.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [decimal.o] Error 1
make: *** [libcxx] Error 2
```

Based on limited results from google, i.e. [this](https://stackoverflow.com/a/43699536), does that mean that im SOL since i cant upgrade to xcode 8 because im running 10.10.5? I'm using [this](https://xcodereleases.com/) as reference, and tbh i dont really want to install xcode anyways as i dont need it... It also seems like im getting all sorts of error for running an old xcode... 

```
$ brew install ant
...
==> Installing dependencies for ant: openjdk
==> Installing ant dependency: openjdk
==> ./configure --disable-warnings-as-errors --with-boot-jdk-jvmargs=-Duser.home=/Users/MrAtheist/Lib
Last 15 lines from /Users/MrAtheist/Library/Logs/Homebrew/openjdk/01.configure:
checking for /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++ option to enable C++11 features... none needed
checking how to run the C preprocessor... /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang -E
checking how to run the C++ preprocessor... /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++ -E
configure: Using clang linker version 253.9 [@(#)PROGRAM:ld  PROJECT:ld64-253.9]
checking for ar... /usr/bin/ar
checking for lipo... /usr/bin/lipo
checking for otool... /usr/bin/otool
checking for install_name_tool... /usr/bin/install_name_tool
checking for metal... [not found]
checking if metal can be run using xcrun... yes, will be using 'xcrun -sdk macosx metal'
checking for metallib... [not found]
checking if metallib can be run using xcrun... no
configure: error: XCode tool 'metallib' neither found in path nor with xcrun
/private/tmp/openjdk-20211010-22597-1liqp36/jdk17u-jdk-17-ga/build/.configure-support/generated-configure.sh: line 84: 5: Bad file descriptor
configure exiting with result code 1
```

I'll keep digging... or i guess the worst case is to install whatever straight from the source.

**Reply by Bo98** [reply]
Ok, great - sounds like brew itself is working now. As for individual formulae, there's little guarantee they will work. They are not tested at all on older macOS but we do accept patches to fix them.

Taking `mpdecimal` for example, we could disable the C++ side of it on older Xcode via `--disable-cxx` (the plain C part still works and is what Python etc use).

For OpenJDK, it might just be the case that it doesn't support older Xcode anymore. You might need to use an older OpenJDK, such as `openjdk@11`. I'm not sure if OpenJDK supports building without `metallib` anymore.

**Reply by MrAtheist** [reply]
> Taking mpdecimal for example, we could disable the C++ side of it on older Xcode via --disable-cxx

Out of curiosity, how would this be done? is it hidden deep within the brew source code? or something else that a nub like me can easily do it?

At some point when im fed up with all of this nonsense, i'll probably just nuke the whole laptop and upgrade to 10.15...

**Reply by astiob** [reply]
You can `brew edit mpdecimal` and add `--disable-cxx` to its `configure` options inside `def install`. However, be warned that this will eventually cause conflicts with `brew update` when the formula is edited in upstream Homebrew.

Another option is to `brew install gcc` or `brew install llvm` and use those compilers to compile the formulæ that fail with your Xcode-provided Clang, like so: `brew install --cc=gcc node` or `brew install --cc=llvm_clang node`. Caveats:

* Installing either `gcc` or `llvm` from source will take a while and require quite a bit of free disk space during the build. Be prepared.
* Try to stick to just one option and avoid mixing them together, because C++ software built with one generally cannot work with the other (except when C++ is used only internally and not exposed). Similarly, you can’t mix C++ software built using brewed GCC and using your system Clang, whether you built it from source or poured it from an old bottle.
* I’m not entirely sure, but I think `--cc=llvm_clang` defaults to the system libc++. If that’s true, then it _can_ be mixed with system-Clang-built things, but on the other hand it can’t build software that needs a newer C++ library (e. g. `mkvtoolnix`). `brew info llvm` offers a suggestion for enabling the newer, custom libc++, but I don’t think it can be readily applied to `brew install`.
* If you choose `gcc`, and _possibly_ if you choose `llvm` with custom libc++, be prepared to rebuild all C++ formulæ when the compiler’s major version changes.
    * With GCC, you may want to use a versioned formula, e. g. `gcc-10`, instead of the cutting-edge `gcc`, and correspondingly use `--cc=gcc-10`, and stick to that version until Homebrew drops it or you find you need a still-newer GCC to compile something; and only then install a newer GCC and start rebuilding everything.
    * With LLVM, if this is relevant at all, I believe Homebrew doesn’t offer an equivalent `--cc=llvm_clang-X` flag, so this isn’t readily possible.
